pkg/scheduler/algorithm/predicates/metadata.go

```diff
@@ -248,22 +240,25 @@ func getTPMapMatchingSpreadConstraints(pod *v1.Pod, nodeInfoMap map[string]*sche
 		nodeTopologyMaps := newTopologyPairsMaps()
 		// nodeInfo.Pods() can be empty; or all pods don't fit
 		for _, existingPod := range nodeInfo.Pods() {
-			ok, err := podMatchesAllSpreadConstraints(existingPod, pod.Namespace, constraints)
-			if err != nil {
-				catchError(err)
-				cancel()
-				return
+			if existingPod.Namespace != pod.Namespace {
+				continue
 			}
-			if ok {
-				for _, constraint := range constraints {
+			podLabelSet := labels.Set(existingPod.Labels)
+			for _, constraint := range constraints {
+				ok, err := podMatchesSpreadConstraint(podLabelSet, constraint)
+				if err != nil {
+					errCh.SendErrorWithCancel(err, cancel)
+					return
+				}
+				if ok {
 					// constraint.TopologyKey is already guaranteed to be present
 					pair := topologyPair{key: constraint.TopologyKey, value: node.Labels[constraint.TopologyKey]}
```

Comment:
You may also want to add that since we restrict a toplogyKey to exist in a single hard constraint, using topolgyKey as the key in this map is good enough to identify which exact constraint this pod is counted against (because otherwise the key to the map should be some sort of an identifier that combines the topologyKey and the label selector)